### PR TITLE
Reduce the number of times to call git command

### PR DESCRIPTION
**Pull request by qinezh**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Common/Git/GitUtility.cs [hunk 1]
```diff
@@ -59,7 +59,31 @@ public static GitRepoInfo GetRepoInfo(string directory)
...
+        public static bool IsGitRoot(string directory)
+        {
+            if (!PathUtility.IsDirectory(directory))
+            {
+                throw new ArgumentException($"{directory} should be a directory");
+            }
+
+            if (!Directory.Exists(directory))
+            {
+                throw new FileNotFoundException($"{directory} can't be found");
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/Git/GitUtility.cs]: DirectoryNotFoundException

File: src/Microsoft.DocAsCode.Common/Git/GitUtility.cs [hunk 2]
```diff
@@ -59,7 +59,31 @@ public static GitRepoInfo GetRepoInfo(string directory)
...
             }
 
-            return Cache.GetOrAdd(directory, GetRepoInfoCore);
+            if (IsGitRoot(directory))
+            {
+                return Cache.GetOrAdd(directory, GetRepoInfoCore);
+            }
+
+            var parentDirInfo = Directory.GetParent(directory);
+            return Cache.GetOrAdd(directory, GetRepoInfo(parentDirInfo?.FullName));
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/Git/GitUtility.cs]: when directory is root directory, exception will throw?

Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/Git/GitUtility.cs]: >GetRepoInfo [](start = 45, length = 11)

This recursion will cause more calling ExistGitCommand(). Extract a GetRepoInfoCore() or put this check somewhere else?

Comment (qinezh) [inline src/Microsoft.DocAsCode.Common/Git/GitUtility.cs]: Yes, `Directory.GetParent` will throw the exception.

---
In reply to: [94093186](https://github.com/dotnet/docfx/pull/1077#discussion_r94093186) [](ancestors = 94093186)

Comment (qinezh) [inline src/Microsoft.DocAsCode.Common/Git/GitUtility.cs]: why are there more calling ExistGitCommand()? I think it's the same as before.  However, I should reduce the number of times to call ExistGitCommand too...

---
In reply to: [94094370](https://github.com/dotnet/docfx/pull/1077#discussion_r94094370) [](ancestors = 94094370)

**Comment by superyyrrzz** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Common/Git/GitUtility.cs [hunk 3]
```diff
@@ -12,6 +12,7 @@ public static class GitUtility
     {
         private static readonly string CommandName = "git";
         private static readonly int GitTimeOut = 1000;
+        private static int _existGitCommand = -1;
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/Git/GitUtility.cs]: >int [](start = 23, length = 3)

bool?

**Comment by vwxyzh** [comment]
:shipit:

**Pull request merged**